dataproc/quickstart/quickstart.go

```diff
@@ -0,0 +1,221 @@
+	createOp, err := clusterClient.CreateCluster(ctx, createReq)
+	if err != nil {
+		log.Fatalf("error submitting the cluster creation request: %v", err)
+	}
+
+	createResp, err := createOp.Wait(ctx)
+	if err != nil {
+		log.Fatalf("error creating the cluster: %v", err)
+	}
+
+	// Output a success message.
+	fmt.Printf("Cluster created successfully: %q", createResp.ClusterName)
+
+	// Create the job client.
+	jobClient, err := dataproc.NewJobControllerClient(ctx, option.WithEndpoint(endpoint))
+
+	// Create the job config.
+	submitJobReq := &dataprocpb.SubmitJobRequest{
+		ProjectId: projectID,
+		Region:    region,
+		Job: &dataprocpb.Job{
+			Placement: &dataprocpb.JobPlacement{
+				ClusterName: clusterName,
+			},
+			TypeJob: &dataprocpb.Job_PysparkJob{
+				PysparkJob: &dataprocpb.PySparkJob{
+					MainPythonFileUri: jobFilePath,
+				},
+			},
+		},
+	}
+
+	submitJobResp, err := jobClient.SubmitJob(ctx, submitJobReq)
+	if err != nil {
+		fmt.Errorf("error submitting job: %v", err)
+	}
+
+	id := submitJobResp.Reference.JobId
+
+	fmt.Printf("Submitted job %q", id)
+
+	// These states all signify that a job has terminated, successfully or not.
+	terminalStates := map[dataprocpb.JobStatus_State]bool{
+		dataprocpb.JobStatus_ERROR:     true,
+		dataprocpb.JobStatus_CANCELLED: true,
+		dataprocpb.JobStatus_DONE:      true,
+	}
+
+	// We can create a timeout such that the job gets cancelled if not in a terminal state after a certain amount of time.
+	timeout := 5 * time.Minute
+	start := time.Now()
+
+	state := submitJobResp.Status.State
+	for !terminalStates[state] {
+		if time.Since(start) > timeout {
+			cancelReq := &dataprocpb.CancelJobRequest{
+				ProjectId: projectID,
+				Region:    region,
+				JobId:     id,
+			}
+
+			_, err := jobClient.CancelJob(ctx, cancelReq)
+			if err != nil {
+				fmt.Errorf("error cancelling job: %v", err)
+			}
+			fmt.Errorf("job %q timed out after %d minutes", id, int64(timeout.Minutes()))
+			break
+		}
+
+		getJobReq := &dataprocpb.GetJobRequest{
+			ProjectId: projectID,
+			Region:    region,
+			JobId:     id,
+		}
+		getJobResp, err := jobClient.GetJob(ctx, getJobReq)
+		if err != nil {
+			fmt.Errorf("error getting job %q with error: %v", id, err)
+			break
+		}
+		state = getJobResp.Status.State
+
+		time.Sleep(1 * time.Second)
+	}
+
+	// Cloud Dataproc job outget gets saved to a GCS bucket allocated to it.
+	getCReq := &dataprocpb.GetClusterRequest{
+		ProjectId:   projectID,
+		Region:      region,
+		ClusterName: clusterName,
+	}
+
+	resp, err := clusterClient.GetCluster(ctx, getCReq)
+	if err != nil {
+		fmt.Errorf("error getting cluster %q: %v", clusterName, err)
+	}
+
+	storageClient, err := storage.NewClient(ctx)
+	if err != nil {
+		fmt.Errorf("error creating storage client: %v", err)
+	}
+
+	obj := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", resp.ClusterUuid, id)
+	reader, err := storageClient.Bucket(resp.Config.ConfigBucket).Object(obj).NewReader(ctx)
```

Comment:
Should we only read the result if the job does not time out? Use `defer` for the delete so we can `return` on timeout and the cluster will still get deleted? Or maybe put it into a function you can call here and in the cancel block?